### DISCUSSION TITLE
3.13 pip install fails

### DISCUSSION THREAD

**Discussion by noblockhit** [pyopencl/pyopencl#801] [Troubleshooting]
**Description**
When i try to install pyopencl with pip on windows 11 or windows 10, it fails and raises this error:
```
Building wheels for collected packages: pyopencl
  Building wheel for pyopencl (pyproject.toml) ... error
  error: subprocess-exited-with-error

  × Building wheel for pyopencl (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [80 lines of output]
      *** scikit-build-core 0.10.7 using CMake 3.29.3 (wheel)
      *** Configuring CMake...
      loading initial cache file C:\Users\bubig\AppData\Local\Temp\tmpl9e7eg40\build\CMakeInit.txt
      -- Building for: Visual Studio 17 2022
      -- Selecting Windows SDK version 10.0.22621.0 to target Windows 10.0.22631.
      -- The CXX compiler identification is MSVC 19.41.34123.0
      -- Detecting CXX compiler ABI info
      -- Detecting CXX compiler ABI info - done
      -- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.41.34120/bin/Hostx64/x64/cl.exe - skipped
      -- Detecting CXX compile features
      -- Detecting CXX compile features - done
      -- CL_INC_DIR not set, trying to guess it from environment variables.
      -- CL_INC_DIR not set, trying to guess it from conda environment.
      -- CL_INC_DIR
      -- CL_LIB_DIR
      -- CL_LIBNAME
      -- Found Python: D:\AA_CODING\python\Projects\AA_MFSIG\StackingCL\Scripts\python.exe (found version "3.13.0") found components: Interpreter Development.Module NumPy
      'D:\AA_CODING\python\Projects\AA_MFSIG\StackingCL\Scripts\python.exe' '-c' 'import nanobind; print(nanobind.__version__)'
      'D:\AA_CODING\python\Projects\AA_MFSIG\StackingCL\Scripts\python.exe' '-c' 'import nanobind; print(nanobind.cmake_dir())'
      -- Looking for CL_VERSION_3_0
      -- Looking for CL_VERSION_3_0 - found
      -- Found OpenCL: C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v12.0/lib/x64/OpenCL.lib (found version "3.0")
      ==============================
      PyOpenCL v2024.3 (non-git) configuration:
        PyOpenCL options: PYOPENCL_TRACE=OFF PYOPENCL_ENABLE_GL=OFF PYOPENCL_USE_SHIPPED_EXT=OFF PYOPENCL_PRETEND_CL_VERSION=
        OpenCL:           C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v12.0/lib/x64/OpenCL.lib [3.0]
        Python:           D:\AA_CODING\python\Projects\AA_MFSIG\StackingCL\Scripts\python.exe [3.13.0]
        Build type:
        C++ compiler:     C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.41.34120/bin/Hostx64/x64/cl.exe [MSVC 19.41.34123.0]
        CMake:            C:/Program Files/CMake/bin/cmake.exe [3.29.3]
        Nanobind:         C:\Users\bubig\AppData\Local\Temp\pip-build-env-s75_6929\overlay\Lib\site-packages\nanobind\cmake [2.2.0]
        Build tool:       C:/Program Files/Microsoft Visual Studio/2022/Community/MSBuild/Current/Bin/amd64/MSBuild.exe
      ==============================
      -- Configuring done (2.7s)
      -- Generating done (0.0s)
      -- Build files have been written to: C:/Users/bubig/AppData/Local/Temp/tmpl9e7eg40/build
      *** Building project with Visual Studio 17 2022...
      MSBuild version 17.11.9+a69bbaaf5 for .NET Framework

      C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppBuild.targets(541,5): warning MSB8029: The Intermediate directory or Output directory cannot reside under the Temporary directory as it could lead to issues with incremental build. [C:\Users\bubig\AppData\Local\Temp\tmpl9e7eg40\build\ZERO_CHECK.vcxproj]
        1>Checking Build System
      C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppBuild.targets(541,5): warning MSB8029: The Intermediate directory or Output directory cannot reside under the Temporary directory as it could lead to issues with incremental build. [C:\Users\bubig\AppData\Local\Temp\tmpl9e7eg40\build\nanobind-static.vcxproj]     
        Building Custom Rule C:/Users/bubig/AppData/Local/Temp/pip-install-3vk6j_i5/pyopencl_b8e68b964c3142aca873a04dbbc62b1d/CMakeLists.txt
        nb_internals.cpp
        nb_func.cpp
        nb_type.cpp
        nb_enum.cpp
        nb_ndarray.cpp
        nb_static_property.cpp
        common.cpp
        error.cpp
        trampoline.cpp
        implicit.cpp
        Generating Code...
        nanobind-static.vcxproj -> C:\Users\bubig\AppData\Local\Temp\tmpl9e7eg40\build\Release\nanobind-static.lib
      C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppBuild.targets(541,5): warning MSB8029: The Intermediate directory or Output directory cannot reside under the Temporary directory as it could lead to issues with incremental build. [C:\Users\bubig\AppData\Local\Temp\tmpl9e7eg40\build\_cl.vcxproj]
        Building Custom Rule C:/Users/bubig/AppData/Local/Temp/pip-install-3vk6j_i5/pyopencl_b8e68b964c3142aca873a04dbbc62b1d/CMakeLists.txt
        wrap_constants.cpp
        wrap_cl.cpp
        wrap_cl_part_1.cpp
        wrap_cl_part_2.cpp
        wrap_mempool.cpp
        bitlog.cpp
      C:\Users\bubig\AppData\Local\Temp\pip-install-3vk6j_i5\pyopencl_b8e68b964c3142aca873a04dbbc62b1d\src\wrap_cl.hpp(70,10): error C1083: Cannot open include file: 'CL/cl.h': No such file or directory [C:\Users\bubig\AppData\Local\Temp\tmpl9e7eg40\build\_cl.vcxproj]
        (compiling source file '../../pip-install-3vk6j_i5/pyopencl_b8e68b964c3142aca873a04dbbc62b1d/src/wrap_cl.cpp')

      C:\Users\bubig\AppData\Local\Temp\pip-install-3vk6j_i5\pyopencl_b8e68b964c3142aca873a04dbbc62b1d\src\wrap_cl.hpp(70,10): error C1083: Cannot open include file: 'CL/cl.h': No such file or directory [C:\Users\bubig\AppData\Local\Temp\tmpl9e7eg40\build\_cl.vcxproj]
        (compiling source file '../../pip-install-3vk6j_i5/pyopencl_b8e68b964c3142aca873a04dbbc62b1d/src/wrap_constants.cpp')

      C:\Users\bubig\AppData\Local\Temp\pip-install-3vk6j_i5\pyopencl_b8e68b964c3142aca873a04dbbc62b1d\src\wrap_cl.hpp(70,10): error C1083: Cannot open include file: 'CL/cl.h': No such file or directory [C:\Users\bubig\AppData\Local\Temp\tmpl9e7eg40\build\_cl.vcxproj]
        (compiling source file '../../pip-install-3vk6j_i5/pyopencl_b8e68b964c3142aca873a04dbbc62b1d/src/wrap_cl_part_1.cpp')

      C:\Users\bubig\AppData\Local\Temp\pip-install-3vk6j_i5\pyopencl_b8e68b964c3142aca873a04dbbc62b1d\src\wrap_cl.hpp(70,10): error C1083: Cannot open include file: 'CL/cl.h': No such file or directory [C:\Users\bubig\AppData\Local\Temp\tmpl9e7eg40\build\_cl.vcxproj]
        (compiling source file '../../pip-install-3vk6j_i5/pyopencl_b8e68b964c3142aca873a04dbbc62b1d/src/wrap_cl_part_2.cpp')

      C:\Users\bubig\AppData\Local\Temp\pip-install-3vk6j_i5\pyopencl_b8e68b964c3142aca873a04dbbc62b1d\src\wrap_cl.hpp(70,10): error C1083: Cannot open include file: 'CL/cl.h': No such file or directory [C:\Users\bubig\AppData\Local\Temp\tmpl9e7eg40\build\_cl.vcxproj]
        (compiling source file '../../pip-install-3vk6j_i5/pyopencl_b8e68b964c3142aca873a04dbbc62b1d/src/wrap_mempool.cpp')


      *** CMake build failed
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for pyopencl
Failed to build pyopencl
ERROR: ERROR: Failed to build installable wheels for some pyproject.toml based projects (pyopencl)
```

**To Reproduce**
1. Install python3.13 (i always installed 3.13t too)
2. select 3.13 or 3.13t as your interpreter
3. install pyopencl using `py -m pip install pyopencl`
4. See error above shown

**Environment (please complete the following information):**
 - OS: Windows 10/11
 - ICD Loader and version: [e.g. ocl-icd 2.3.1]
 - ICD and version: [e.g. pocl 1.8]
 - CPU/GPU: Ryzen 7 3800x with 2060OC / Intel i7 7th gen with 940mx / intel 640
  - Python version: 3.13 / 3.13t
  - PyOpenCL version: latest (2024.3)
  
  
  Is pyopencl even supported on 3.13? I couldn't find any concrete information on that, neither in forums, pypi nor here on github.

**Comment by inducer** [comment]
From your install log:

> `Cannot open include file: 'CL/cl.h'`

You're missing OpenCL headers. If you have them installed somewhere that CMake's [OpenCL finder](https://cmake.org/cmake/help/latest/module/FindOpenCL.html) doesn't look by default, [you can specify the spot manually](https://documen.tician.de/pyopencl/misc.html#installing-from-source).

3.13 is supported, has been for a while.

**Reply by noblockhit** [reply]
First of all, thank you for your quick answer @inducer, i had other responsibilities and i only just now got to try it.
So, i have installed opencl headers, compiled it for opencl-icd-loader too, because i needed the lib file later down the road. The documentation is pretty good btw, chat gpt led me down the wrong path, should've looked at the readme, rtfm ik. then, as i didnt find a proper .whl for python 3.13 here: https://pypi.org/project/pyopencl/#files i decided to compile it from source, which went well enough. the installation also appeared to work. Then, when i tried to import pyopencl like this:
```
Python 3.13.0 (tags/v3.13.0:60403a5, Oct  7 2024, 09:38:07) [MSC v.1941 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyopencl
C:\Users\bubig\AppData\Roaming\Python\Python313\site-packages\pytools\persistent_dict.py:52: RecommendedHashNotFoundWarning: Unable to import recommended hash 'siphash24.siphash13', falling back to 'hashlib.sha256'. Run 'python3 -m pip install siphash24' to install the recommended hash.
  warn("Unable to import recommended hash 'siphash24.siphash13', "
Traceback (most recent call last):
  File "<python-input-0>", line 1, in <module>
    import pyopencl
  File "C:\Users\bubig\AppData\Roaming\Python\Python313\site-packages\pyopencl\__init__.py", line 32, in <module>
    import pyopencl.cltypes
  File "C:\Users\bubig\AppData\Roaming\Python\Python313\site-packages\pyopencl\cltypes.py", line 25, in <module>
    from pyopencl.tools import get_or_register_dtype
  File "C:\Users\bubig\AppData\Roaming\Python\Python313\site-packages\pyopencl\tools.py", line 140, in <module>
    from pyopencl.compyte.dtypes import (  # noqa: F401
    ...<4 lines>...
    )
ModuleNotFoundError: No module named 'pyopencl.compyte'
>>> 
```
i got the error above too. Then i tried to install the hash thingy (you never know) which failed and i didn't look further into it, i just remember the error message being something, that cl cant compile something. Fair enough, i tried to google the error and found this:
https://github.com/inducer/pyopencl/commit/65d3c3228f943b607c5a084c52d990b9c66bfb71

Either I'm completely on the wrong path here, or there is no way to install it for 3.13. 
More strangely, what i initially wanted to point out is: for python3.10 and 3.11 (i didnt try 3.12) the installation went flawlessly just by pip, without any other shenanigans like manually having to reference dependencies, it just worked using pip install pyopencl. why is this not the case now? I have the real strong impression that something is wrong with 3.13 in general. Tkinter doesn't work, and installing some other modules (like the hash thingy but not limited to) don't work either (For the record, my own, terribly designed and very simple shitty module (sbNative) works just fine in the install for me on both my desktop and laptop platform on 3.13). Any advice? do i just wait for 3.13.1 where hopefully everything will work just fine like it should? What do i do or can i help by any means?

**Reply by inducer** [reply]
Please describe how you install pyopencl. I ask because the following steps provide a perfectly functional installation on my system, as expected:
```
$ python3.13 -m venv pyopencl-install
$ source pyopencl-install/bin/activate
$ pip install pyopencl
```

**Reply by noblockhit** [reply]
![image](https://github.com/user-attachments/assets/3d59632a-d3aa-46bf-aeb8-9ce7161fa182)

And this is my install command:
py -m pip install -v . --config-settings='cmake.args=-DCL_INC_DIR=D:\CODING\drivers\OpenCL-Headers;-DCL_LIB_DIR=D:\CODING\drivers\OpenCL-ICD-Loader-main'

the file searched in the screenshot and also mentioned above compyte indeed doesn't exist.

**Reply by alexfikl** [reply]
From your traceback, it seems like you're using the free-threaded "no-GIL" version of Python (it has version `3.13t` where the `t` denotes the no-GIL version). Can you try installing the standard version instead and using that?

I don't think `pyopencl` works with the no-GIL version, since the bindings make use the GIL all over the place. I'm not sure that's the issue, but it's worth a try..

**Reply by noblockhit** [reply]
I did also try that, which failed. But i have found a (terrible) solution: after installing it and getting the error that pyopencl.compyte doesn't exist, I took a brief look into the 3.12 install on my machine and simply copied the folder compyte to the install location of 3.13t. I ran a test program that chatgpt gave me, which worked just fine. To test all the functionality i need, i'd like to run my own program but tkinter is broken and some other modules won't install too. I guess I'll have to wait for the necessary fixes in order to properly test everything, but this hotfix solved the issue with pyopencl for now. I'd still recommend you to investigate why installing from pypi directly doesn't work and you have to build three projects on your own (OpenCL-Headers, OpenCL-ICD-Loader and pyopencl). I am very certain that something is wrong with python 3.13t and/or pip on 3.13t and it might not even be your fault. For example: the temporary folder that is supposed to be created (C:\Users\bubig\AppData\Local\Temp\tmpl9e7eg40\ in my case) never seemed to exist. I'm not entirely sure and i didn't system log it but I never got to see it myself in that location.
Thank you for your time though.

PS: pyopencl also worked when being forced to free thread (-Xgil=0 option)
